### PR TITLE
Do not wrap platform locks in `sys_common`

### PR DESCRIPTION
**Pull request by joboet**
Now that `Mutex` and `RwLock` are `const`, `StaticMutex` and `StaticRwLock` are no longer necessary. Removing them results in a extra few allocations on platforms which require boxed locks (currently some UNIXes and SGX), but it makes the codebase quite a lot cleaner. With the lock improvements tracked in #93740, these allocations will become unnecessary.

The rest of the lock abstractions in `sys_common` can be removed by modifying the lock implementations in `sys` so that they are `const` and movable on all platforms. This entails a slight bit of code duplication (both SGX and UNIX need to wrap their locks in `LazyBox`, for now), but reduces the usage of `unsafe` quite considerably.

The `Condvar` check is merged into the `pthread_condvar` implementation, making `sys_common::condvar` redundant. The check is otherwise only performed on SGX and Hermit, which however do not rely on the check for soundness. I therefore removed it on these platforms, as it will be made redundant by the aforementioned lock improvements anyway. 

Additionally, this PR makes the condition variable on Hermit movable by lazily initializing it in place instead of lazily allocating and initializing.

@rustbot label +T-libs, +T-libs-api
r? @m-ou-se

### CODE REVIEW

**Comment by m-ou-se** [comment]
Always good to see a PR that removes more code than it adds. ^^

It's quite a large PR though, affecting a lot of different things, so it'd be good if you could split it up a bit. Then if this accidentally causes some issues later, it'll be easier to track down.

- It seems to me it should be easy to split off the changes that make all `MovableThing = Thing` on all platforms. (One PR for all platforms, or one PR each; whatever you prefer.) Then we can merge that first, making the remaining change less complicated.

- The removal of `imp::requires_synchronized_create()` is mostly unrelated to this change, and should be possible to do separately in an independent PR.

- The changes to `stdout()` don't only remove the `Pin`, but also move the lazy initialization of the buffer from the `stdout()` call to the first `write()` call. Is there any benefit in doing it that way? If so, it'd be good to put that in a separate PR.

File: library/std/src/sys/unix/thread_parker.rs [hunk 1]
```diff
@@ -56,7 +56,8 @@ unsafe fn wait_timeout(
         target_os = "macos",
         target_os = "ios",
         target_os = "watchos",
-        target_os = "espidf"
+        target_os = "espidf",
+        target_os = "horizon",
```
Comment (m-ou-se) [inline library/std/src/sys/unix/thread_parker.rs:60]: Why did this change? This seems unrelated to this PR.

**Comment by joboet** [comment]
I got a bit too carried away with the changes... :wink:

This PR is now split in #100576, #100579, #100581 and #100583, the fifth part will be removing the `sys_common` wrappers, but I'll wait with that until the others are merged.
